(/integrations/webhooks/sync-data) update guide (#1655)
clerk/clerk-docs: docs/integrations/webhooks/sync-data.mdx

---
title: Sync Clerk data to your app with webhooks
description: Learn how to sync Clerk data to your app with webhooks.
---

<TutorialHero
  beforeYouStart={[

      {
      title: "A Clerk + Next.js app is required.",
      link: "/docs/quickstarts/nextjs",
      icon: "nextjs",
    },
    {
      title: "A ngrok account is required.",
      link: "https://dashboard.ngrok.com/signup",
      icon: "user-circle",
    }
  ]}
>
  - Set up ngrok
  - Set up a webhook endpoint
  - Create the webhook
  - Get type inference for your webhook events
  - Test the webhook
  - Configure your production instance
</TutorialHero>

The recommended way to sync Clerk data to your app is through webhooks.

In this guide, you'll set up a webhook in your Next.js app to listen for the `user.created` event, create an endpoint in the Clerk Dashboard, build a handler, and test it locally using ngrok and the Clerk Dashboard.

> Clerk offers many user events, but three key events include:

Reviewer: ```suggestion
Clerk offers many events, but three key events include:
```
`user.created`, `user.updated` and `user.deleted` are the only user events. The others are for things like Organizations, Organization Memberships, Organization Invitations, Sessions, etc

- `user.created`: Triggers when a new user registers in the app or is created via the Clerk Dashboard or Backend API. Listening to this event allows the initial insertion of user information in your database.
- `user.updated`: Triggers when user information is updated via Clerk components, the Clerk Dashboard, or Backend API. Listening to this event keeps data synced between Clerk and your external database. It is recommended to only sync what you need to simplify this process.

> - `user.deleted`: Triggers when a user deletes their account, or their account is removed via the Clerk Dashboard or Backend API. Listening to this event allows you to delete the user from your database or add a `deleted: true` flag.

Reviewer 2: what was the reasoning behind adding these three and context to them?

Reviewer: Prior to review, this guide used just `user.updated` which is a weird spot for a customer to start with. Most customers would want to start with `user.created`. Its not required, but I think mentioning them is valuable. This was the original review comment -> https://github.com/clerk/clerk-docs/pull/1655#discussion_r1823433618

  1. In the **Subscribe to events** section, scroll down and select `user.created`.
  1. Select **Create**. You'll be redirected to your endpoint's settings page. Keep this page open.

  ### Add your Signing Secret to `.env.local`

  To verify the webhook payload, you'll need your endpoint's **Signing Secret**. Since you don't want this secret exposed in your codebase, store it as an environment variable in your `.env.local` file during local development.

  1. On the endpoint's settings page, copy the **Signing Secret**.
  1. In your project's root directory, open or create an `.env.local` file, which should already include your Clerk API keys. Assign your **Signing Secret** to `SIGNING_SECRET`. The file should resemble:

  ```env {{ filename: '.env.local' }}
  NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY={{pub_key}}
  CLERK_SECRET_KEY={{secret}}
  SIGNING_SECRET=whsec_123
  ```

  ### Set the webhook route as public in your Middleware

  Incoming webhook events have no auth information. They come from an external source and will not be signed in or signed out. To allow access, the route must be public.

>   Incoming webhook events contain no authentication information. They come from an external source and won't be signed in or signed out, so the route must be public to allow access.

Reviewer: These are repeated. I believe Alexis left a longer note about this as well.

  Incoming webhook events have no auth information. They come from an external source and will not be signed in or signed out. To allow access, the route must be public.

  Incoming webhook events contain no authentication information. They come from an external source and won't be signed in or signed out, so the route must be public to allow access.

>   To configure your Middleware, refer to the [clerkMiddleware() guide](docs/references/nextjs/clerk-middleware).

Reviewer 2: looks like you repeat information here. 
also the sentence ends wtih "so the route must be public to allow access". and then the next sentence goes into configuring your middleware.

there should be some relational information that ties these two concepts together:
"so the route must be public to allow access. If you are using `clerkMiddleware()`, you will need to ensure that the `/api/webhooks(.*)` route is public. For more information on configuring routes with `clerkMiddleware()`, see [the reference](/docs/references/nextjs/clerk-middleware).

also don't forget that clerkMiddleware() always needs to be in code snippets. even if it's in a link

  ### Install `svix`

  Clerk uses [`svix`](https://www.npmjs.com/package/svix) to deliver webhooks, so you'll need the package to verify the webhook signature. Run the following command in your terminal to install it:

>   <CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>

Reviewer 2: deprecated property - let's do our best to remove it whenever we see it

```suggestion
  <CodeBlockTabs options={["npm", "yarn", "pnpm"]}>
```

  ### Create the endpoint

  Set up a Route Handler that uses `svix` to verify the incoming Clerk webhook and process the payload.

  For this guide, the payload will be logged to the console. In a real app, you'd use the payload to trigger an action. For example, if listening for the `user.created` event, you might perform a database `create` or `upsert` to add the user's details to the user's table.

>   If the route handler returns a [2xx code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses), the event will be marked as successful, and retries will stop. If the route handler returns a [4xx](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) or [5xx code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#server_error_responses), or no code at all, the webhook event will be [retried](/docs/integrations/webhooks/overview#retry).
> 
>   In summary, a 2xx (success) code stops retries, while any other response is considered a failure and will trigger retries.

Reviewer 2: I don't think there's a need for a "summary" sentence.

```suggestion
  If the route handler returns a [4xx](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) or [5xx code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#server_error_responses), or no code at all, the webhook event will be [retried](/docs/integrations/webhooks/overview#retry). If the route handler returns a [2xx code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses), the event will be marked as successful, and retries will stop.
```

    ```
  </CodeBlockTabs>

  ### Create the endpoint

  Set up a Route Handler that uses `svix` to verify the webhook signature and process the payload.

  For this guide, the payload will be logged to the console. In a real app, you'd use the payload to trigger an action. For example, if listening for the `user.updated` event, you might perform a database `update` or `upsert` to update the user's details.

>   The webhook must return either an error code or a success code, such as `200` or `201`. An error code will appear in the Clerk Dashboard log and trigger a [retry](/docs/integrations/webhooks/overview#retry). A success code will stop further retries and show a success status in the Clerk Dashboard log.

Reviewer: ```suggestion
  The webhook must return either an error code such as `405` or `403` or a success code, such as `200` or `201`. An error code will appear in the Clerk Dashboard log and trigger a [retry](/docs/integrations/webhooks/overview#retry). A success code will stop further retries and show a success status in the Clerk Dashboard log.
```

Maybe link to https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses ?

  > [!NOTE]

>   > The following Route Handler can be used for any webhook event you choose to listen to, not just `user.updated`.

Reviewer: ```suggestion
  > The following Route Handler will listen to all webhook event types. The next section will demonstrate how to listen to a specific event and also get type inference for that event type.
```

  <Tabs type="framework" items={["Next.js", "Express"]}>
    <Tab>
      <CodeBlockTabs type="router" options={["App Router"]}>
        ```ts {{ filename: 'app/api/webhooks/route.ts' }}
        import { Webhook } from 'svix'
        import { headers } from 'next/headers'
        import { WebhookEvent } from '@clerk/nextjs/server'

        export async function POST(req: Request) {
          const SIGNING_SECRET = process.env.SIGNING_SECRET

          if (!SIGNING_SECRET) {

>             throw new Error('Please add SIGNING_SECRET from Clerk Dashboard to .env or .env.local')

Reviewer: ```suggestion
            throw new Error('Error: Please add SIGNING_SECRET from Clerk Dashboard to .env or .env.local')
```
Just to be consistent with other errors.

    ```
  </CodeBlockTabs>

  ### Create the endpoint

  Set up a Route Handler that uses `svix` to verify the webhook and process the payload.

  For this guide, the payload will be logged to the console. In a real app, you'd use the payload to trigger an action. For example, if listening for the `user.created` event, you might perform a database `update` or `upsert` to update the user's details.

>   The webhook must return either an [error code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) such as `405` or `403` or a [success code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses), such as `200` or `201`. An error code will appear in the Clerk Dashboard log and trigger a [retry](/docs/integrations/webhooks/overview#retry). A success code will stop further retries and show a success status in the Clerk Dashboard log.

Reviewer: This should probably be more like:

If the route handler returns a 2xx code that will mark the even as successful and stop retries.
If the route handler returns a 4xx or 5xx code or no code at all, then the webhook event will be retried.

Basically a 2xx/success code stops retries and anything else does not and is considered a failure.

            }) as WebhookEvent
          } catch (err) {

>             console.error('Error verifying webhook:', err)

Reviewer: ```suggestion
            console.error('Error: Could not verify webhook:', err)
```
Just to be consistent with other errors.

>             return new Response('Verification error', {

Reviewer: ```suggestion
            return new Response('Error: Verification error', {
```
Just to be consistent.

              status: 400,
            })
          }

>           // Do something with the payload

Reviewer 2: ```suggestion
          // Do something with payload
```

>           // Log payload to console

Reviewer 2: ```suggestion
          // For this guide, log payload to console
```

          const { id } = evt.data
          const eventType = evt.type
          console.log(`Received webhook with ID ${id} and event type of ${eventType}`)
          console.log('Webhook payload:', body)

          return new Response('Webhook received', { status: 200 })
        }
        ```
      </CodeBlockTabs>
    </Tab>

    <Tab>
      ```ts {{ filename: 'index.ts' }}
      app.post(
        '/api/webhooks',
        // This is a generic method to parse the contents of the payload.
        // Depending on the framework, packages, and configuration, this may be
        // different or not required.
        bodyParser.raw({ type: 'application/json' }),

        async (req, res) => {
          const SIGNING_SECRET = process.env.SIGNING_SECRET

          if (!SIGNING_SECRET) {

>             throw new Error('Please add SIGNING_SECRET from Clerk Dashboard to .env')

Reviewer: ```suggestion
            throw new Error('Error: Please add SIGNING_SECRET from Clerk Dashboard to .env')
```

Just to be consistent with other errors.

          }

          // Create new Svix instance with secret
          const wh = new Webhook(SIGNING_SECRET)

          // Get headers and body
          const headers = req.headers
          const payload = req.body

          // Get Svix headers for verification
          const svix_id = headers['svix-id']
          const svix_timestamp = headers['svix-timestamp']
          const svix_signature = headers['svix-signature']

          // If there are no headers, error out
          if (!svix_id || !svix_timestamp || !svix_signature) {
            return void res.status(400).json({
              success: false,
              message: 'Error: Missing svix headers',
            })
          }

          let evt

          // Attempt to verify the incoming webhook
          // If successful, the payload will be available from 'evt'

>           // If  verification fails, error out and return error code

Reviewer 2: ```suggestion
          // If verification fails, error out and return error code
```

          try {
            evt = wh.verify(payload, {
              'svix-id': svix_id as string,
              'svix-timestamp': svix_timestamp as string,
              'svix-signature': svix_signature as string,
            })
          } catch (err) {

>             console.log('Error verifying webhook:', err.message)

Reviewer: ```suggestion
            console.log('Error: Could not verify webhook:', err.message)
```
Just to be consistent with other errors.

            return void res.status(400).json({
              success: false,
              message: err.message,
            })
          }

>           // Do something with the payload

Reviewer 2: ```suggestion
          // Do something with payload
```

>           // Log payload to console

Reviewer 2: ```suggestion
          // For this guide, log payload to console
```

    </Tab>
  </Tabs>

  ### Narrow to a webhook event for type inference

  `WebhookEvent` encompasses all possible webhook types. Narrow down the event type for accurate typing for specific events.

>   In the following example, the `if` statement narrows the type to `user.created`, enabling type-safe access to evt.data with autocompletion.

Reviewer 2: the guide is about listening for the user.updated event

```suggestion
  In the following example, the `if` statement narrows the type to `user.updated`, enabling type-safe access to evt.data with autocompletion.
```

>   In the following example, the `if` statement narrows the type to `user.updated`, enabling type-safe access to evt.data with autocompletion.

Reviewer: Possibly change this depending on the note above about the `user.*` events.

  ```ts {{ filename: 'app/api/webhooks/route.ts', del: [1, 2], ins: [[4, 6]] }}
  console.log(`Received webhook with ID ${id} and event type of ${eventType}`)
  console.log('Webhook payload:', body)

>   if (evt.type === 'user.created') {

Reviewer 2: ```suggestion
  if (evt.type === 'user.updated') {
```

  - `SMSMessageJSON`
  - `UserJSON`

  ### Test the webhook

  1. Start your Next.js server.
  1. In your endpoint's settings page in the Clerk Dashboard, select the **Testing** tab.
  1. In the **Select event** dropdown, select `user.created`.
  1. Select **Send Example**.
  1. In the **Message Attempts** section, confirm that the event is labeled with `Succeeded`.

  #### Handling failed messages

  1. In the **Message Attempts** section, select the event labeled with `Failed`.
  1. Scroll down to the **Webhook Attempts** section.
  1. Toggle the arrow next to the **Status** column.
  1. Review the error. Solutions vary by error type. For more information, refer to the [Debug your webhooks](/docs/integrations/webhooks/debug-your-webhooks) guide.

  ### Trigger the webhook

  To trigger the `user.created` event, you can do either one of the following:

  1. Edit your user in the Clerk Dashboard.
  1. Select the `<UserProfile />` component in your app to edit your profile.

>   Once you have updated a user, you should be able to see the webhook's payload logged to your terminal. You can also check the Clerk Dashboard to see the webhook attempt, the same way you did when [testing the webhook](#test-the-webhook).

Reviewer: Maybe not worth changing, but technically this guide doesn't actually update the user. Maybe just this?
```suggestion
  You should be able to see the webhook's payload logged to your terminal. You can also check the Clerk Dashboard to see the webhook attempt, the same way you did when [testing the webhook](#test-the-webhook).
```

</Steps>

### Configure your production instance

1. When you're ready to deploy your app to production, follow [the guide on deploying your Clerk app to production](/docs/deployments/overview).
1. Create your production webhook by following the steps in the previous [Set up a webhook endpoint](#set-up-a-webhook-endpoint) section. In the **Endpoint URL** field, instead of pasting the ngrok URL, paste your production app URL.
1. After you've set up your webhook endpoint, you'll be redirected to your endpoint's settings page. Copy the **Signing Secret**.

> 1. On your hosting platform, update your environment variables by adding **Signing Secret** with the key of `SIGNING_SECRET`.

Reviewer: ```suggestion
1. On your hosting platform, update your environment variables on your hosting platform by adding **Signing Secret** with the key of `SIGNING_SECRET`.
```

  - `SMSMessageJSON`
  - `UserJSON`

  ### Test the webhook

  1. Start your Next.js server.
  1. In your endpoint's settings page in the Clerk Dashboard, select the **Testing** tab.
  1. In the **Select event** dropdown, select `user.updated`.
  1. Select **Send Example**.
  1. In the **Message Attempts** section, confirm that the event is labeled with `Succeeded`.

  #### Handling failed messages

  1. In the **Message Attempts** section, select the event labeled with `Failed`.
  1. Scroll down to the **Webhook Attempts** section.
  1. Toggle the arrow next to the **Status** column.
  1. Review the error. Solutions vary by error type. For more information, refer to the [Debug your webhooks](/docs/integrations/webhooks/debug-your-webhooks) guide.

  ### Trigger the webhook

  To trigger the `user.updated` event, you can do either one of the following:

  1. Edit your user in the Clerk Dashboard.
  1. Select the `<UserProfile />` component in your app to edit your profile.

>   Once you have updated a user, you should be able to see the webhook's payload logged to your terminal. You can also check the Clerk Dashboard to see the webhook attempt, the same way you did when [testing the webhook](#test-your-webhook).

Reviewer 2: ```suggestion
  Once you have updated a user, you should be able to see the webhook's payload logged to your terminal. You can also check the Clerk Dashboard to see the webhook attempt, the same way you did when [testing the webhook](#test-the-webhook).
```

> </Steps>

Reviewer: Another callout to potentially update this depending on the above

### Configure your production instance

1. When you're ready to deploy your app to production, follow [the guide on deploying your Clerk app to production](/docs/deployments/overview).
1. Create your production webhook by following the steps in the previous [Set up a webhook endpoint](#set-up-a-webhook-endpoint) section. In the **Endpoint URL** field, instead of pasting the ngrok URL, paste your production app URL.

> 1. After you've set up your webhook endpoint, you'll be redirected to your endpoint's settings page. Copy the **Signing Secret** and update your `.env.local` file with the new **Signing Secret**.

Reviewer: ```suggestion
1. After you've set up your webhook endpoint, you'll be redirected to your endpoint's settings page. Copy the **Signing Secret**.
```

Don't update `.env.local` with the production Signing Secret. That would only go in the hosting platform. `.env.local` would retain the initial/development Signing Secret.
